core/chains/evm/gas/arbitrum_estimator.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/pkg/errors"
+	"golang.org/x/exp/slices"
+
+	evmclient "github.com/smartcontractkit/chainlink/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+type ArbConfig interface {
+	EvmGasLimitMax() uint32
+}
+
+type ethClient interface {
+	CallContract(ctx context.Context, msg ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
+}
+
+// arbitrumEstimator is an Estimator which extends l2SuggestedPriceEstimator to use getPricesInArbGas() for gas limit estimation.
+type arbitrumEstimator struct {
+	utils.StartStopOnce
+
+	cfg ArbConfig
+
+	Estimator // *l2SuggestedPriceEstimator
+
+	client     ethClient
+	pollPeriod time.Duration
+	logger     logger.Logger
+
+	getPricesInArbGasMu sync.RWMutex
+	perL2Tx             uint32
+	perL1CalldataUnit   uint32
+
+	chForceRefetch chan (chan struct{})
+	chInitialised  chan struct{}
+	chStop         chan struct{}
+	chDone         chan struct{}
+}
+
+func NewArbitrumEstimator(lggr logger.Logger, cfg ArbConfig, rpcClient rpcClient, ethClient ethClient) Estimator {
+	lggr = lggr.Named("ArbitrumEstimator")
+	return &arbitrumEstimator{
+		cfg:            cfg,
+		Estimator:      NewL2SuggestedPriceEstimator(lggr, rpcClient),
+		client:         ethClient,
+		pollPeriod:     10 * time.Second,
+		logger:         lggr,
+		chForceRefetch: make(chan (chan struct{})),
+		chInitialised:  make(chan struct{}),
+		chStop:         make(chan struct{}),
+		chDone:         make(chan struct{}),
+	}
+}
+
+func (a *arbitrumEstimator) Start(ctx context.Context) error {
+	return a.StartOnce("ArbitrumEstimator", func() error {
+		if err := a.Estimator.Start(ctx); err != nil {
+			return errors.Wrap(err, "failed to start gas price estimator")
+		}
+		go a.run()
+		<-a.chInitialised
+		return nil
+	})
+}
+func (a *arbitrumEstimator) Close() error {
+	return a.StopOnce("ArbitrumEstimator", func() (err error) {
+		close(a.chStop)
+		err = errors.Wrap(a.Estimator.Close(), "failed to stop gas price estimator")
+		<-a.chDone
+		return
+	})
+}
+
+// GetLegacyGas estimates both the gas price and the gas limit.
+func (a *arbitrumEstimator) GetLegacyGas(calldata []byte, l2GasLimit uint32, maxGasPriceWei *big.Int, opts ...Opt) (gasPrice *big.Int, chainSpecificGasLimit uint32, err error) {
+	gasPrice, _, err = a.Estimator.GetLegacyGas(calldata, l2GasLimit, maxGasPriceWei, opts...)
+	if err != nil {
+		return
+	}
+	ok := a.IfStarted(func() {
+		if slices.Contains(opts, OptForceRefetch) {
+			ch := make(chan struct{})
+			select {
+			case a.chForceRefetch <- ch:
+			case <-a.chStop:
+				err = errors.New("estimator stopped")
+				return
+			}
+			select {
+			case <-ch:
+			case <-a.chStop:
+				err = errors.New("estimator stopped")
+				return
+			}
+		}
+		perL2Tx, perL1CalldataUnit := a.getPricesInArbGas()
+		// TODO is it ok to return lsGasLimit if unitialized?
+		chainSpecificGasLimit = l2GasLimit + perL2Tx + uint32(len(calldata))*perL1CalldataUnit
```

Comment:
This needs comments or at least a link to some docs somewhere

Reply:
I am not aware of official docs for this formula. :detective:

Reply:
I think just add comments on the struct level, or function level, on what they do.